# Data Wrangling on WeRateDogs Tweets
by Shan Jafri

In a world where data is abundent, gathering and cleaning data is necessary for any data scientist to extract valuable insights. The focus of this project is to gather data from multiple sources, assess dirty data and then clean the data. The last step of the project is to show several insights from the clean data set.

# Gather
In this project, data was gathered from three sources:

The WeRateDogs twitter archive that was provided by Udacity as a csv file.
The image prediction file that was hosted by Udacity serves and downloaded programmatically using request libary
Using twitters API and python's tweepy library, any missing information from WeRateDogs twiiter archive such as retweet counts and favorite counts. The tweet IDs in the WeRateDogs Twitter archive will be use to query the Twitter API for each tweet's JSON data using Python's Tweepy library and store each tweet's entire set of JSON data in a file called tweet_json.txt file. Each tweet's JSON data should be written to its own line. Then we will read this .txt file line by line into a pandas DataFrame with (at minimum) tweet ID, retweet count, and favorite count.

** Let's begin gathering data!**

In [1]:
# Import the libaries needed for this project
import pandas as pd 
import numpy as np 
import requests
import os
import tweepy
import json
from timeit import default_timer as timer

In [2]:
# Open twitter archive enhanced and read file in pandas
twitter_archive = pd.read_csv('twitter-archive-enhanced.csv')
# view the file 
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

** The file twitter archive has 17 columns and 2356 entries.**

In [3]:
# Request image prediction file from Udacity Server and read in Pandas
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
response = requests.get(url)

with open (url.split('/')[-1], mode = 'wb') as file:
    file.write(response.content)

image_prediction = pd.read_csv('image-predictions.tsv', sep = '\t')

In [4]:
# View image prediction file 
image_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


#### Extracting data using Twitters API 

In [5]:
# token and keys
consumer_key = '6GF8J1A1tZQODiu5Hch3O8Mru'
consumer_secret = 'dAQWLTIeGCYWHWCBUHaOR7QA3KVdfLaHsqky54lecigFkt2K6V'
access_token = '1047201527423815680-cnOWUtr4zWolbcFrkWabmzrvWK4BZ8'
access_token_secret = '8wMkUZH6pUQnwQU0IyX3XPwctA30yAf81tTFDZNV7Cn8P' 

# Authenticating user access
auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_token_secret)

# Contruct the API instance
api = tweepy.API(auth, 
                 parser=tweepy.parsers.JSONParser(), # parse the results to JSON objects 
                 wait_on_rate_limit=True, # Automatically wait for rate limits to replenish
                 wait_on_rate_limit_notify=True) # Print notification when Tweepy is waiting to replenish

In [6]:
# Store the results
tweet_dic = []
# Store tweet error
error = []

start = timer() # Calculating the length of time
# Get tweet JSON objects for all tweets in twitter archive data set
for tweet in twitter_archive['tweet_id']:
    try:
        tweet_data = api.get_status(tweet, tweet_mode = 'extended')
        # Get JSON Data of retweet and favorite count
        retweets = tweet_data['retweet_count']
        favorites = tweet_data['favorite_count']
        
        tweet_dic.append({'tweet_id': tweet,
                          'retweet': retweets,
                          'favorite': favorites})
        
    # Catch the exceptions of the TweepError
    except Exception as e:
        print(str(tweet)+ " _ " + str(e))
        error.append(tweet)

Rate limit reached. Sleeping for: 97


888202515573088257 _ [{'code': 144, 'message': 'No status found with that ID.'}]
873697596434513921 _ [{'code': 144, 'message': 'No status found with that ID.'}]
872668790621863937 _ [{'code': 144, 'message': 'No status found with that ID.'}]
869988702071779329 _ [{'code': 144, 'message': 'No status found with that ID.'}]
866816280283807744 _ [{'code': 144, 'message': 'No status found with that ID.'}]
861769973181624320 _ [{'code': 144, 'message': 'No status found with that ID.'}]
845459076796616705 _ [{'code': 144, 'message': 'No status found with that ID.'}]
842892208864923648 _ [{'code': 144, 'message': 'No status found with that ID.'}]
837012587749474308 _ [{'code': 144, 'message': 'No status found with that ID.'}]
827228250799742977 _ [{'code': 144, 'message': 'No status found with that ID.'}]
802247111496568832 _ [{'code': 144, 'message': 'No status found with that ID.'}]
775096608509886464 _ [{'code': 144, 'message': 'No status found with that ID.'}]
771004394259247104 _ [{'code

Rate limit reached. Sleeping for: 623


754011816964026368 _ [{'code': 144, 'message': 'No status found with that ID.'}]


Rate limit reached. Sleeping for: 634


In [13]:
# lengh of the result
print("The lengh of the result", len(tweet_dic))
# The tweet_id of the errors
print("The lengh of the errors", len(error))

The lengh of the result 2341
The lengh of the errors 15


In [32]:
# Store tweet error
second_error = []

start = timer() # Calculating the length of time
# Get tweet JSON objects for all tweets in twitter archive data set
for er in error:
    try:
        tweet_data = api.get_status(tweet, tweet_mode = 'extended')
        # Get JSON Data of retweet and favorite count
        retweets = tweet_data['retweet_count']
        favorites = tweet_data['favorite_count']
        
        tweet_dic.append({'tweet_id': tweet,
                          'retweet': retweets,
                          'favorite': favorites})
        
    # Catch the exceptions of the TweepError
    except Exception:
        print(str(tweet)+ " _ " + str(e))
        second_error.append(er)
end = timer()
print(end - start)

4.528745600000548


In [33]:
# write json object into text file
with open('tweet_json.txt', 'w') as outfile:
    json.dump(tweet_dic, outfile)

# Turn JSON file into a DataFrame
json_df = pd.read_json('tweet_json.txt')

In [34]:
json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2446 entries, 0 to 2445
Data columns (total 3 columns):
favorite    2446 non-null int64
retweet     2446 non-null int64
tweet_id    2446 non-null int64
dtypes: int64(3)
memory usage: 57.4 KB


### Gathering Summary

The first step of data wrangling process is complete. 
- data is downloaded from exisiting file
- data is requested and downloaded from a Udacity server 
- data is query from titter's API
- All three data is gathered and imported into Jupyter Notebook


## Assess

The gathered data has to be assessed for the range of the cleaning required. When assessing the gathered data, the types of cleaning is broken into two categories, tidy and dirty data issues.  

#### Json dataframe

In [35]:
# Check number of columns and entries
json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2446 entries, 0 to 2445
Data columns (total 3 columns):
favorite    2446 non-null int64
retweet     2446 non-null int64
tweet_id    2446 non-null int64
dtypes: int64(3)
memory usage: 57.4 KB


In [46]:
# Visual inspection of data set
json_df

favorite  retweet            tweet_id
0        38296     8406  892420643555336193
1        32826     6198  892177421306343426
2        24726     4100  891815181378084864
3        41643     8536  891689557279858688
4        39814     9253  891327558926688256
5        19987     3077  891087950875897856
6        11702     2042  890971913173991426
7        64655    18642  890729181411237888
8        27477     4215  890609185150312448
9        31527     7299  890240255349198849
10       30292     7243  890006608113172480
11       27444     4915  889880896479866881
12       47529     9938  889665388333682689
13       26844     4489  889638837579907072
14       14906     2212  889531135344209921
15       24985     5316  889278841981685760
16       28731     4445  888917238123831296
17       25278     4246  888804989199671297
18       19640     3508  888554962724278272
19       21486     3451  888078434458587136
20       29794     5322  887705289381826560
21       45701    11522  887517139158093824
22       68270    17964  887473957103951883
23       33268    10280  887343217045368832
24       30181     5890  887101392804085760
25       34731     7681  886983233522544640
26       11905     3248  886736880519319552
27       22155     4408  886680336477933568
28       20933     3163  886366144734445568
29         116        4  886267009285017600
...        ...      ...                 ...
2416      2549      508  666020888022790149
2417      2549      508  666020888022790149
2418      2549      508  666020888022790149
2419      2549      508  666020888022790149
2420      2549      508  666020888022790149
2421      2549      508  666020888022790149
2422      2549      508  666020888022790149
2423      2549      508  666020888022790149
2424      2549      508  666020888022790149
2425      2549      508  666020888022790149
2426      2549      508  666020888022790149
2427      2549      508  666020888022790149
2428      2549      508  666020888022790149
2429      2549      508  666020888022790149
2430      2549      508  666020888022790149
2431      2549      508  666020888022790149
2432      2549      508  666020888022790149
2433      2549      508  666020888022790149
2434      2549      508  666020888022790149
2435      2549      508  666020888022790149
2436      2549      508  666020888022790149
2437      2549      508  666020888022790149
2438      2549      508  666020888022790149
2439      2549      508  666020888022790149
2440      2549      508  666020888022790149
2441      2549      508  666020888022790149
2442      2549      508  666020888022790149
2443      2549      508  666020888022790149
2444      2549      508  666020888022790149
2445      2549      508  666020888022790149

[2446 rows x 3 columns]

In [37]:
# Check number of duplicate tweets
json_df.duplicated(subset= ['tweet_id','favorite','retweet']).sum()

105

#### Twitter archive dataframe

In [38]:
# Number of columns and duplicates
twitter_archive.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [45]:
# Visual inspection of data set
twitter_archive

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     892420643555336193                    NaN                  NaN   
1     892177421306343426                    NaN                  NaN   
2     891815181378084864                    NaN                  NaN   
3     891689557279858688                    NaN                  NaN   
4     891327558926688256                    NaN                  NaN   
5     891087950875897856                    NaN                  NaN   
6     890971913173991426                    NaN                  NaN   
7     890729181411237888                    NaN                  NaN   
8     890609185150312448                    NaN                  NaN   
9     890240255349198849                    NaN                  NaN   
10    890006608113172480                    NaN                  NaN   
11    889880896479866881                    NaN                  NaN   
12    889665388333682689                    NaN                  NaN   
13    889638837579907072                    NaN                  NaN   
14    889531135344209921                    NaN                  NaN   
15    889278841981685760                    NaN                  NaN   
16    888917238123831296                    NaN                  NaN   
17    888804989199671297                    NaN                  NaN   
18    888554962724278272                    NaN                  NaN   
19    888202515573088257                    NaN                  NaN   
20    888078434458587136                    NaN                  NaN   
21    887705289381826560                    NaN                  NaN   
22    887517139158093824                    NaN                  NaN   
23    887473957103951883                    NaN                  NaN   
24    887343217045368832                    NaN                  NaN   
25    887101392804085760                    NaN                  NaN   
26    886983233522544640                    NaN                  NaN   
27    886736880519319552                    NaN                  NaN   
28    886680336477933568                    NaN                  NaN   
29    886366144734445568                    NaN                  NaN   
...                  ...                    ...                  ...   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2329  666373753744588802                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2332  666345417576210432                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2336  666273097616637952                    NaN                  NaN   
2337  666268910803644416                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2339  666102155909144576                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2343  666073100786774016                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2345  666063827256086533                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN                  NaN   
2348  666055525042405380                    NaN                  NaN   
2349  666051853826850816                    NaN         

In [42]:
# number of duplicate tweets
twitter_archive.duplicated(['tweet_id']).sum()

0

In [47]:
# Name columns shows incorrect dog names
twitter_archive.name

0        Phineas
1          Tilly
2         Archie
3          Darla
4       Franklin
5           None
6            Jax
7           None
8           Zoey
9         Cassie
10          Koda
11         Bruno
12          None
13           Ted
14        Stuart
15        Oliver
16           Jim
17          Zeke
18       Ralphus
19        Canela
20        Gerald
21       Jeffrey
22          such
23        Canela
24          None
25          None
26          Maya
27        Mingus
28         Derek
29        Roscoe
          ...   
2326       quite
2327           a
2328        None
2329        None
2330        None
2331        None
2332        None
2333          an
2334           a
2335          an
2336        None
2337        None
2338        None
2339        None
2340        None
2341        None
2342        None
2343        None
2344        None
2345         the
2346         the
2347           a
2348           a
2349          an
2350           a
2351        None
2352           a
2353          

### Image Prediction DataFrame

In [54]:
image_prediction

tweet_id                                            jpg_url  \
0     666020888022790149    https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1     666029285002620928    https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2     666033412701032449    https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3     666044226329800704    https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4     666049248165822465    https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   
5     666050758794694657    https://pbs.twimg.com/media/CT5Jof1WUAEuVxN.jpg   
6     666051853826850816    https://pbs.twimg.com/media/CT5KoJ1WoAAJash.jpg   
7     666055525042405380    https://pbs.twimg.com/media/CT5N9tpXIAAifs1.jpg   
8     666057090499244032    https://pbs.twimg.com/media/CT5PY90WoAAQGLo.jpg   
9     666058600524156928    https://pbs.twimg.com/media/CT5Qw94XAAA_2dP.jpg   
10    666063827256086533    https://pbs.twimg.com/media/CT5Vg_wXIAAXfnj.jpg   
11    666071193221509120    https://pbs.twimg.com/media/CT5cN_3WEAAlOoZ.jpg   
12    666073100786774016    https://pbs.twimg.com/media/CT5d9DZXAAALcwe.jpg   
13    666082916733198337    https://pbs.twimg.com/media/CT5m4VGWEAAtKc8.jpg   
14    666094000022159362    https://pbs.twimg.com/media/CT5w9gUW4AAsBNN.jpg   
15    666099513787052032    https://pbs.twimg.com/media/CT51-JJUEAA6hV8.jpg   
16    666102155909144576    https://pbs.twimg.com/media/CT54YGiWUAEZnoK.jpg   
17    666104133288665088    https://pbs.twimg.com/media/CT56LSZWoAAlJj2.jpg   
18    666268910803644416    https://pbs.twimg.com/media/CT8QCd1WEAADXws.jpg   
19    666273097616637952    https://pbs.twimg.com/media/CT8T1mtUwAA3aqm.jpg   
20    666287406224695296    https://pbs.twimg.com/media/CT8g3BpUEAAuFjg.jpg   
21    666293911632134144    https://pbs.twimg.com/media/CT8mx7KW4AEQu8N.jpg   
22    666337882303524864    https://pbs.twimg.com/media/CT9OwFIWEAMuRje.jpg   
23    666345417576210432    https://pbs.twimg.com/media/CT9Vn7PWoAA_ZCM.jpg   
24    666353288456101888    https://pbs.twimg.com/media/CT9cx0tUEAAhNN_.jpg   
25    666362758909284353    https://pbs.twimg.com/media/CT9lXGsUcAAyUFt.jpg   
26    666373753744588802    https://pbs.twimg.com/media/CT9vZEYWUAAlZ05.jpg   
27    666396247373291520    https://pbs.twimg.com/media/CT-D2ZHWIAA3gK1.jpg   
28    666407126856765440    https://pbs.twimg.com/media/CT-NvwmW4AAugGZ.jpg   
29    666411507551481857    https://pbs.twimg.com/media/CT-RugiWIAELEaq.jpg   
...                  ...                                                ...   
2045  886366144734445568    https://pbs.twimg.com/media/DE0BTnQUwAApKEH.jpg   
2046  886680336477933568    https://pbs.twimg.com/media/DE4fEDzWAAAyHMM.jpg   
2047  886736880519319552    https://pbs.twimg.com/media/DE5Se8FXcAAJFx4.jpg   
2048  886983233522544640    https://pbs.twimg.com/media/DE8yicJW0AAAvBJ.jpg   
2049  887101392804085760    https://pbs.twimg.com/media/DE-eAq6UwAA-jaE.jpg   
2050  887343217045368832  https://pbs.twimg.com/ext_tw_video_thumb/88734...   
2051  887473957103951883    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2052  887517139158093824  https://pbs.twimg.com/ext_tw_video_thumb/88751...   
2053  887705289381826560    https://pbs.twimg.com/media/DFHDQBbXgAEqY7t.jpg   
2054  888078434458587136    https://pbs.twimg.com/media/DFMWn56WsAAkA7B.jpg   
2055  888202515573088257    https://pbs.twimg.com/media/DFDw2tyUQAAAFke.jpg   
2056  888554962724278272    https://pbs.twimg.com/media/DFTH_O-UQAACu20.jpg   
2057  888804989199671297    https://pbs.twimg.com/media/DFWra-3VYAA2piG.jpg   
2058  888917238123831296    https://pbs.twimg.com/media/DFYRgsOUQAARGhO.jpg   
2059  889278841981685760  https://pbs.twimg.com/ext_tw_video_thumb/88927...   
2060  889531135344209921    https://pbs.twimg.com/media/DFg_2PVW0AEHN3p.jpg   
2061  889638837579907072    https://pbs.twimg.com/media/DFihzFfXsAYGDPR.jpg   
2062  889665388333682689    https://pbs.twimg.com/media/DFi579UWsAAatzw.jpg   
2063  889880896479866881    https://pbs.twimg.com/media/DFl99B1WsA

In [52]:
image_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [64]:
# Search for duplicates
image_prediction.duplicated('tweet_id').sum()

0

In [66]:
# Search for duplicates
image_prediction.duplicated(subset = ['jpg_url']).sum()

66

In [107]:
image_prediction.isna().sum()

tweet_id    0
jpg_url     0
img_num     0
p1          0
p1_conf     0
p1_dog      0
p2          0
p2_conf     0
p2_dog      0
p3          0
p3_conf     0
p3_dog      0
dtype: int64

### Quality
  ##### Twitter_archive


- tweet_id should be a string instead of an integer
- timestamp column should be datetime data type
- 'Name' columns have incorrect names like 'a', 'such', 'None'
- rename columns 'text' = 'tweet_text', 'name' = 'dog_name', 'source' = 'source_application'
- in_reply_to_status_id/user_id, retweeted_status_id/user_id should be srings instead of float
- ratings_numerator/denominator, retweets, and favorites should be integers instead of floats
   
   
  ##### Image_prediction
- Tweet_id should be a string instead of an integer
- Missing values from images dataset only 2075 entires
- Some tweet_ids have the same jpg_url
- Source columncan be cleaned

  ##### json dataframe
- remove any duplicate tweet_id's


### Tidiness

- Merge all data tables together
- Dog description in columns instead of rows
- Too many columns in image prediciton could condense
- Only interested in unique ratings, no retweets
- Drop unnecessary columns and NaN rows
- retweeted_status_id/user_id should be dropped

Cleaning tidy data issues first is often the best way to start. When structral data issues are solved, cleaning quality data issues become simpler.

** Let's begin cleaning**

## Clean

In [703]:
# First step is to create a copy 
json_clean = json_df.copy()
twitter_clean = twitter_archive.copy()
image_clean = image_prediction.copy()

### Tidiness Data Cleaning

### Tidy Data Issue 1
 - Merge all data tables together

In [704]:
# The first step is going to merge all the data sets together.
twitter_master = pd.merge(twitter_clean, json_clean, how = 'left', on = 'tweet_id')
twitter_master = pd.merge(twitter_master, image_clean, how = 'left', on = 'tweet_id')

### Test Data Issue 1

In [705]:
# view twitter_master dataframe
twitter_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2461 entries, 0 to 2460
Data columns (total 30 columns):
tweet_id                      2461 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2461 non-null object
source                        2461 non-null object
text                          2461 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2402 non-null object
rating_numerator              2461 non-null int64
rating_denominator            2461 non-null int64
name                          2461 non-null object
doggo                         2461 non-null object
floofer                       2461 non-null object
pupper                        2461 non-null object
puppo                         2461 non-null object
favorite                      24

### Tidy Data Issue 2

- Only interested in unique ratings, no retweets
- Drop unnecessary columns and NaN rows

In [706]:
twitter_master.isnull().sum()

tweet_id                         0
in_reply_to_status_id         2383
in_reply_to_user_id           2383
timestamp                        0
source                           0
text                             0
retweeted_status_id           2280
retweeted_status_user_id      2280
retweeted_status_timestamp    2280
expanded_urls                   59
rating_numerator                 0
rating_denominator               0
name                             0
doggo                            0
floofer                          0
pupper                           0
puppo                            0
favorite                        15
retweet                         15
jpg_url                        281
img_num                        281
p1                             281
p1_conf                        281
p1_dog                         281
p2                             281
p2_conf                        281
p2_dog                         281
p3                             281
p3_conf             

In [707]:
# Row entries that have retweets need to be dropped
twitter_master = twitter_master[pd.isnull(twitter_master['retweeted_status_id'])]
# Drop duplicate tweet_id
twitter_master.drop_duplicates(inplace = True)
# Drop any retweet related columns
twitter_master.drop(columns = ['retweeted_status_id', 'retweeted_status_user_id',
                                 'retweeted_status_timestamp'], inplace = True)

What entries can we drop from the twitter master dataframe? There are 181 NaN values in the image prediction portion of the twitter master dataframe. There's also NaN values in the retweet and favorite column We can drop these rows.

In [708]:
twitter_master.dropna(subset = ['jpg_url', 'img_num', 'p1','p1_conf','p1_dog',
                                'p2','p2_conf','p2_dog','p3','p3_conf','p3_dog',
                               'favorite','retweet'], inplace = True)

### Test Data Issue 2

In [709]:
twitter_master.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1993 entries, 0 to 2355
Data columns (total 27 columns):
tweet_id                 1993 non-null int64
in_reply_to_status_id    23 non-null float64
in_reply_to_user_id      23 non-null float64
timestamp                1993 non-null object
source                   1993 non-null object
text                     1993 non-null object
expanded_urls            1993 non-null object
rating_numerator         1993 non-null int64
rating_denominator       1993 non-null int64
name                     1993 non-null object
doggo                    1993 non-null object
floofer                  1993 non-null object
pupper                   1993 non-null object
puppo                    1993 non-null object
favorite                 1993 non-null float64
retweet                  1993 non-null float64
jpg_url                  1993 non-null object
img_num                  1993 non-null float64
p1                       1993 non-null object
p1_conf              

### Tidy Data Issue 3

- The four dog description each have there own columns when they should be condensed into a row using the melt function

In [710]:
# Create columns list for the melt function
columns_condensed = ['doggo', 'floofer', 'pupper', 'puppo']
columns_unchanged = []
for x in twitter_master.columns:
    if x not in columns_condensed:
        columns_unchanged.append(x)

In [711]:
twitter_master = pd.melt(twitter_master, id_vars = columns_unchanged, value_vars = columns_condensed,
       var_name = 'column_names', value_name = 'dog_description')


In [712]:
twitter_master = twitter_master.sort_values('dog_description').drop_duplicates(
    subset = ['tweet_id'], keep = 'last')
twitter_master.drop(columns = ['column_names'], inplace = True)
twitter_master.reset_index(drop = True, inplace = True)

### Test Data Issue 3

In [713]:
twitter_master.dog_description.value_counts()

None       1687
pupper      212
doggo        63
puppo        23
floofer       8
Name: dog_description, dtype: int64

All the columns for the dog description are in one column. The majority of entries for dog description are 'None'. 

### Tidy Data Issue 4

- dog prediction that are have labeled as True starting with the first prediction and move down to second and third predictions are kept and put into 'predicted_dog' column. The algorithm's confidence is stored in the 'confidence_algore' column.

In [714]:
predicted_dog = []
confidence_algore = []
def something(data):
    if data['p1_dog'] == True:
        predicted_dog.append(data['p1'])
        confidence_algore.append(data['p1_conf'])
    elif data['p2_dog'] == True:
        predicted_dog.append(data['p2'])
        confidence_algore.append(data['p2_conf'])
    elif data['p3_dog'] == True:
        predicted_dog.append(data['p3'])
        confidence_algore.append(data['p3_conf'])
    else:
        predicted_dog.append('NaN')
        confidence_algore.append(0)

In [715]:
twitter_master.loc[:, 'p1': 'p3_dog'].apply(something, axis = 1)
twitter_master['predicted_dog'] = np.array(predicted_dog)
twitter_master['confidence_algore'] = np.array(confidence_interval)

In [716]:
twitter_master.drop(columns = twitter_master.loc[:,'img_num':'p3_dog'].columns, inplace = True)

### Test Data Issue 4

In [717]:
twitter_master

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0     667405339315146752                    NaN                  NaN   
1     667435689202614272                    NaN                  NaN   
2     667437278097252352                    NaN                  NaN   
3     667443425659232256                    NaN                  NaN   
4     667453023279554560                    NaN                  NaN   
5     667455448082227200                    NaN                  NaN   
6     667470559035432960                    NaN                  NaN   
7     667491009379606528                    NaN                  NaN   
8     667495797102141441                    NaN                  NaN   
9     667502640335572993                    NaN                  NaN   
10    667509364010450944                    NaN                  NaN   
11    667517642048163840                    NaN                  NaN   
12    667524857454854144                    NaN                  NaN   
13    667530908589760512                    NaN                  NaN   
14    667534815156183040                    NaN                  NaN   
15    667538891197542400                    NaN                  NaN   
16    667544320556335104                    NaN                  NaN   
17    667546741521195010                    NaN                  NaN   
18    667549055577362432                    NaN                  NaN   
19    667724302356258817                    NaN                  NaN   
20    667728196545200128                    NaN                  NaN   
21    667766675769573376                    NaN                  NaN   
22    667773195014021121                    NaN                  NaN   
23    667782464991965184                    NaN                  NaN   
24    667793409583771648                    NaN                  NaN   
25    667801013445750784                    NaN                  NaN   
26    667806454573760512                    NaN                  NaN   
27    667393430834667520                    NaN                  NaN   
28    667832474953625600                    NaN                  NaN   
29    667369227918143488                    NaN                  NaN   
...                  ...                    ...                  ...   
1963  690015576308211712                    NaN                  NaN   
1964  822462944365645825                    NaN                  NaN   
1965  831315979191906304                    NaN                  NaN   
1966  675006312288268288                    NaN                  NaN   
1967  765371061932261376                    NaN                  NaN   
1968  701545186879471618                    NaN                  NaN   
1969  675898130735476737                    NaN                  NaN   
1970  855851453814013952                    NaN                  NaN   
1971  793195938047070209                    NaN                  NaN   
1972  825026590719483904                    NaN                  NaN   
1973  889531135344209921                    NaN                  NaN   
1974  889665388333682689                    NaN                  NaN   
1975  825535076884762624                    NaN                  NaN   
1976  743253157753532416                    NaN                  NaN   
1977  744995568523612160                    NaN                  NaN   
1978  859607811541651456                    NaN                  NaN   
1979  802239329049477120                    NaN                  NaN   
1980  780931614150983680                    NaN                  NaN   
1981  803773340896923648                    NaN                  NaN   
1982  790946055508652032                    NaN                  NaN   
1983  756275833623502848                    NaN                  NaN   
1984  878776093423087618                    NaN                  NaN   
1985  867421006826221569                    NaN                  NaN   
1986  752519690950500352                    NaN         

## Quality Data Cleaning

### Quality Data Issue 1
No operations will be done on 'tweet_id' ,'in_reply_to_status_id', 'in_reply_to_user_id' columns and should be converted into strings. The 'favorite' and 'retweets' columns should be converted into integers. The timestamp column are easier to manipulate as datetime data types.

- Convert Columns to the appropriate data types

In [718]:
# Change columns from integers/floats to strings
twitter_master = twitter_master.astype({'tweet_id': 'str', 'in_reply_to_status_id': 'str',
                                       'in_reply_to_user_id': 'str'})
# Change columns from floats to integers
twitter_master = twitter_master.astype({'favorite': 'int32', 'retweet':'int32'})
# Change date to datetime
twitter_master['timestamp'] = pd.to_datetime(twitter_master['timestamp'])

### Test Data Issue 1

In [719]:
twitter_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1993 entries, 0 to 1992
Data columns (total 16 columns):
tweet_id                 1993 non-null object
in_reply_to_status_id    1993 non-null object
in_reply_to_user_id      1993 non-null object
timestamp                1993 non-null datetime64[ns]
source                   1993 non-null object
text                     1993 non-null object
expanded_urls            1993 non-null object
rating_numerator         1993 non-null int64
rating_denominator       1993 non-null int64
name                     1993 non-null object
favorite                 1993 non-null int32
retweet                  1993 non-null int32
jpg_url                  1993 non-null object
dog_description          1993 non-null object
predicted_dog            1993 non-null object
confidence_algore        1993 non-null float64
dtypes: datetime64[ns](1), float64(1), int32(2), int64(2), object(10)
memory usage: 233.6+ KB


### Quality Data Issue 2

There are certain entries under 'name' columns that are incorrect dog names. Certain dog names are labeled as either 'None', 'is', 'O', 'Such', etc., even when the dog name is presented in the text portion of the tweet. 
- Change dog names that are incorrect

In [720]:
# Create a function that evaluates the fist letter of the name as lower case or 'None'
def islower(x):
    if x[0].islower() or x == 'None':
        return True
    else:
        return False


In [721]:
# Created a different table with dog names that potentially need to be corrected
fix_dogname = twitter_master[twitter_master.name.apply(islower)]

There are 644 rows that have either 'None' or a name starting with a lower case letter. What are all the names? 

In [722]:
# Get a value count for all the different names
twitter_master[twitter_master.name.apply(islower)].name.value_counts()

None            546
a                55
the               7
an                6
one               4
very              4
just              3
quite             3
getting           2
my                1
actually          1
by                1
this              1
infuriating       1
space             1
not               1
unacceptable      1
officially        1
his               1
all               1
such              1
light             1
incredibly        1
Name: name, dtype: int64

There are no entries that have any potential dog names in the new dataframe. Using regex, any potential dog names can be extracted into a new data table and then merged back into the original dataframe.

In [723]:
# Create a dataframe with all the corrections
correction = fix_dogname['text'].str.extract(r'that\s([D-Z]\w+)')
correction['second'] = fix_dogname['text'].str.extract(r'name| named\s(\w*\s*[A-Z]\w+)', expand = False)
correction['third'] = fix_dogname['text'].str.extract(r'\whis\sis\s([A-Z]\w+)', expand = False)
correction.rename( columns = {0 : 'first'} , inplace = True)



In [724]:
# Drop rows that do not have any of correction in it and turn NaN values into empty strings
correction.dropna(how = 'all', inplace = True)
correction.replace(np.nan, '', inplace = True)
# Create a final column combining all the corrections from the three columns
correction['name'] = correction['first'] + correction['second'] + correction['third']
correction.head(3)

first   second third     name
6            Kohl           Kohl
10        Tickles        Tickles
15          Klint          Klint

In [725]:
# any entry that is lowercase becomes 'None'
fix_dogname.loc[:, 'name'] = 'None'
fix_dogname.name.value_counts()

/Users/Erudition/anaconda3/lib/python3.6/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


None    644
Name: name, dtype: int64

In [726]:
# update the twitter_master dataframe with correction['final'] column
fix_dogname.update(correction['name'])
fix_dogname.name.value_counts()

/Users/Erudition/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py:5082: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[col] = expressions.where(mask, this, that)


None         615
Jacob          1
Octaviath      1
Kip            1
Berta          1
Cherokee       1
Leroi          1
Spork          1
Johm           1
Cheryl         1
Alfonso        1
Bretagne       1
Kohl           1
Chuk           1
Yoshi          1
Toby           1
Tickles        1
Zeus           1
Pippa          1
Blue           1
Klint          1
Rufus          1
Alphred        1
Jake           1
Alfredo        1
Wylie          1
Jessiga        1
Pepe           1
Guss           1
Hemry          1
Name: name, dtype: int64

In [727]:
# update the twitter_master dataframe with correction['final'] column
twitter_master.update(fix_dogname)

### Test Data Issue 2

In [728]:
# All the names in the dataset
twitter_master.name.value_counts().index.tolist()

['None',
 'Charlie',
 'Cooper',
 'Oliver',
 'Lucy',
 'Tucker',
 'Penny',
 'Winston',
 'Sadie',
 'Toby',
 'Daisy',
 'Lola',
 'Jax',
 'Bo',
 'Koda',
 'Bella',
 'Stanley',
 'Milo',
 'Rusty',
 'Chester',
 'Leo',
 'Dave',
 'Scout',
 'Louis',
 'Buddy',
 'Bailey',
 'Oscar',
 'Clark',
 'Reggie',
 'Sophie',
 'Bear',
 'Oakley',
 'Clarence',
 'Maggie',
 'George',
 'Finn',
 'Chip',
 'Jack',
 'Brody',
 'Dexter',
 'Alfie',
 'Archie',
 'Gary',
 'Gus',
 'Phil',
 'Jerry',
 'Bentley',
 'Winnie',
 'Larry',
 'Duke',
 'Derek',
 'Scooter',
 'Cassie',
 'Rosie',
 'Waffles',
 'Calvin',
 'Jeffrey',
 'Kyle',
 'Shadow',
 'Mia',
 'Max',
 'Wilson',
 'Moose',
 'Loki',
 'Olive',
 'Gerald',
 'Beau',
 'Zeke',
 'Walter',
 'Zoey',
 'Gizmo',
 'Reese',
 'Sunny',
 'Steven',
 'Otis',
 'Malcolm',
 'Sebastian',
 'Wallace',
 'Rufus',
 'Hank',
 'Pippa',
 'Wyatt',
 'Jimothy',
 'Boomer',
 'Sammy',
 'Louie',
 'Samson',
 'Nala',
 'Lily',
 'Vincent',
 'Riley',
 'Bruce',
 'Earl',
 'Ruby',
 'Ellie',
 'Benji',
 'Luca',
 'Jiminy',
 'Pick

### Quality Data Issue 3

Source column can be cleaned to only indicate source instead of entire link

In [731]:
# Clean the source column
twitter_master['source'] = twitter_master['source'].apply(lambda x: re.findall(r'>(.*)<', x)[0])


IndexError: list index out of range

### Test Data Issue 3

In [730]:
twitter_master.head(3)

tweet_id in_reply_to_status_id in_reply_to_user_id  \
0  667405339315146752                   nan                 nan   
1  667435689202614272                   nan                 nan   
2  667437278097252352                   nan                 nan   

            timestamp              source  \
0 2015-11-19 18:13:27  Twitter for iPhone   
1 2015-11-19 20:14:03  Twitter for iPhone   
2 2015-11-19 20:20:22  Twitter for iPhone   

                                                text  \
0  This is Biden. Biden just tripped... 7/10 http...   
1            Ermergerd 12/10 https://t.co/PQni2sjPsm   
2  Never seen this breed before. Very pointy pup....   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/667405339...               7.0   
1  https://twitter.com/dog_rates/status/667435689...              12.0   
2  https://twitter.com/dog_rates/status/667437278...              10.0   

   rating_denominator   name  favorite  retweet  \
0                10.0  Biden     473.0    222.0   
1                10.0   None     308.0     85.0   
2                10.0   None     453.0    244.0   

                                           jpg_url dog_description  \
0  https://pbs.twimg.com/media/CUMZnmhUEAEbtis.jpg            None   
1  https://pbs.twimg.com/media/CUM1OHCW4AEgGSi.jpg            None   
2  https://pbs.twimg.com/media/CUM2qWaWoAUZ06L.jpg            None   

   predicted_dog  confidence_algore  
0  Saint_Bernard           0.381377  
1     Rottweiler           0.999091  
2            NaN           0.000000

### Quality Data Issue 4 

The in_repy_to_status_id and in_reply_to_user_id columns are not necessary for any analysis and can be dropped.

In [732]:
twitter_master.drop(['in_reply_to_status_id', 'in_reply_to_user_id'], 1, inplace = True)

### Test Data Issue 4

In [735]:
twitter_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1993 entries, 0 to 1992
Data columns (total 14 columns):
tweet_id              1993 non-null object
timestamp             1993 non-null datetime64[ns]
source                1993 non-null object
text                  1993 non-null object
expanded_urls         1993 non-null object
rating_numerator      1993 non-null float64
rating_denominator    1993 non-null float64
name                  1993 non-null object
favorite              1993 non-null float64
retweet               1993 non-null float64
jpg_url               1993 non-null object
dog_description       1993 non-null object
predicted_dog         1993 non-null object
confidence_algore     1993 non-null float64
dtypes: datetime64[ns](1), float64(5), object(8)
memory usage: 218.1+ KB


###  Quality Data Issue 5

Rename columns for better description

In [740]:
twitter_master.rename(columns = {'timestamp': 'tweet_date', 'source': 'tweet_source', 'text': 'tweet_text', 
                                        'expanded_urls': 'tweet_url','name': 'dog_name', 'jpg_url': 'tweet_picture', 
                                        'favorite': 'tweet_favorites', 'retweet': 'tweet_retweets',
                                        'prediction_algorithm' : 'dog_breed', 'confidence_algore': 'algorithm_confidence'}, inplace = True)

### Test Data Issue 5

In [741]:
twitter_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1993 entries, 0 to 1992
Data columns (total 14 columns):
tweet_id                1993 non-null object
tweet_date              1993 non-null datetime64[ns]
tweet_source            1993 non-null object
tweet_text              1993 non-null object
tweet_url               1993 non-null object
rating_numerator        1993 non-null float64
rating_denominator      1993 non-null float64
dog_name                1993 non-null object
tweet_favorites         1993 non-null float64
tweet_retweets          1993 non-null float64
tweet_picture           1993 non-null object
dog_description         1993 non-null object
predicted_dog           1993 non-null object
algorithm_confidence    1993 non-null float64
dtypes: datetime64[ns](1), float64(5), object(8)
memory usage: 218.1+ KB


## Storing Data

In [742]:
twitter_master.to_csv('twitter_archive_master.csv', index=False, encoding = 'utf-8')
twitter_master = pd.read_csv('twitter_archive_master.csv')
twitter_master.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1993 entries, 0 to 1992
Data columns (total 14 columns):
tweet_id                1993 non-null int64
tweet_date              1993 non-null object
tweet_source            1993 non-null object
tweet_text              1993 non-null object
tweet_url               1993 non-null object
rating_numerator        1993 non-null float64
rating_denominator      1993 non-null float64
dog_name                1993 non-null object
tweet_favorites         1993 non-null float64
tweet_retweets          1993 non-null float64
tweet_picture           1993 non-null object
dog_description         1993 non-null object
predicted_dog           1685 non-null object
algorithm_confidence    1993 non-null float64
dtypes: float64(5), int64(1), object(8)
memory usage: 218.1+ KB
